query/logicalplan/optimize.go

```diff
@@ -158,6 +163,36 @@ func (p *FilterPushDown) optimize(plan *LogicalPlan, exprs []Expr) {
 	}
 }
 
+// The FilterExchange optimizer adds an exchange step to parallelize filtering
+// TODO more descriptive comments.
+type FilterExchange struct{}
+
+func (p *FilterExchange) Optimize(plan *LogicalPlan) *LogicalPlan {
+	p.optimize(plan, nil)
+	return plan
+}
+
+func (p *FilterExchange) optimize(plan *LogicalPlan, exprs []Expr) {
+	if plan.Filter != nil {
+		numWorkers := runtime.NumCPU() - 1
+		if numWorkers <= 0 {
+			numWorkers = 1
+		}
+		exchangePlan := LogicalPlan{
+			Exchange: &Exchange{
+				Parallelism:  numWorkers,
```

Comment:
Should the degree of parallelism really be decided in the logical plan? To me, it feels like this should happen while building of the physical plan.